lib/mongodb.js

```diff
@@ -1996,17 +1991,50 @@ function optimizedFindOrCreate(model, filter, data, options, callback) {
  * @param {Object} data The data that might contain a decimal property
  * @param {Object} props The model property definitions
  */
-function convertDecimalProps(data, props) {
-  if (debug.enabled) debug('convertDecimalProps props: ', util.inspect(props));
-  for (const p in props) {
-    const prop = props[p];
-    const isDecimal = data[p] && prop && prop.mongodb &&
-      prop.mongodb.dataType &&
-      prop.mongodb.dataType.toLowerCase() === 'decimal128';
+function convertDecimalProps(data, propDef) {
+  if (propDef == null) return data;
+
+  if (Array.isArray(data)) {
+    const arrType = getArrayItemDef(propDef);
+    if (arrType) {
+      data.forEach(function(elem, i) {
+        data[i] = convertDecimalProps(elem, arrType);
+      });
+      if (debug.enabled) debug('convertDecimalProps converted array: ', util.inspect(data));
+    };
+  } else if (!!data && typeof data === 'object') {
+    // !!data: skips executing the code when data is `null`
+    const ownData = Object.getOwnPropertyNames(data);
+    ownData.forEach(function(k) {
+      data[k] = convertDecimalProps(data[k], getObjectDef(propDef, k));
+    });
+    if (debug.enabled) debug('convertDecimalProps converted object: ', util.inspect(data));
+  } else {
+    const isDecimal = propDef && propDef.mongodb &&
+    propDef.mongodb.dataType &&
+    propDef.mongodb.dataType.toLowerCase() === 'decimal128';
     if (isDecimal) {
-      data[p] = Decimal128.fromString(data[p]);
-      debug('convertDecimalProps decimal value: ', data[p]);
+      data = Decimal128.fromString(data);
+      debug('convertDecimalProps decimal value: ', data);
```

Comment:
Let's add `if (debug.enabled)` condition here too for consistency.

Reply:
good catch.